ext/php8/handlers_curl.c

```diff
@@ -79,54 +76,44 @@ static void dd_ch_duplicate_headers(zval *ch_orig, zval *ch_new) {
+    if (dd_headers && (dd_header_array = zend_hash_index_find_ptr(dd_headers, Z_OBJ_HANDLE_P(ch)))) {
+        ZVAL_ARR(&headers, zend_array_dup(dd_header_array));
+    } else {
+        array_init(&headers);
     }
 
-    if (!headers) {
-        zend_hash_init(retval, 0, NULL, ZVAL_PTR_DTOR, 0);
+    int sampling_priority;
+    if (ddtrace_fetch_prioritySampling_from_root(&sampling_priority)) {
+        add_next_index_str(&headers, zend_strpprintf(0, "x-datadog-sampling-priority: %d", sampling_priority));
+    }
+    if (DDTRACE_G(trace_id)) {
+        add_next_index_str(&headers, zend_strpprintf(0, "x-datadog-trace-id: %" PRIu64, (DDTRACE_G(trace_id))));
+    }
+    if (DDTRACE_G(span_ids_top)) {
+        add_next_index_str(&headers,
+                           zend_strpprintf(0, "x-datadog-parent-id: %" PRIu64, (DDTRACE_G(span_ids_top)->id)));
+    }
+    if (DDTRACE_G(dd_origin)) {
+        add_next_index_str(&headers, zend_strpprintf(0, "x-datadog-origin: %s", ZSTR_VAL(DDTRACE_G(dd_origin))));
     }
-    ZVAL_ARR(arg, retval);
-}
 
-static void dd_free_headers_arg(zval *arg) { zend_array_destroy(Z_ARRVAL_P(arg)); }
+    zend_function *setopt_fn = zend_hash_str_find_ptr(EG(function_table), ZEND_STRL("curl_setopt"));
+    zend_constant *http_header_const = zend_hash_str_find_ptr(EG(zend_constants), ZEND_STRL("CURLOPT_HTTPHEADER"));
 
-static void dd_inject_distributed_tracing_headers(zval *ch) {
-    if (dd_inject_func == NULL) {
-        dd_inject_func = zend_string_init(ZEND_STRL("ddtrace\\bridge\\curl_inject_distributed_headers"), 0);
-    }
-    if (zend_hash_exists(EG(function_table), dd_inject_func)) {
-        zend_function **fn_proxy = &dd_curl_inject_fn_proxy;
-        zval retval = {.u1.type_info = IS_UNDEF};
-
-        zval headers;
-        dd_init_headers_arg(&headers, ch);
-
-        ddtrace_sandbox_backup backup = ddtrace_sandbox_begin();
-        dd_should_save_headers = false;  // Don't save our own HTTP headers
-        // Arg 0: CurlHandle $ch
-        // Arg 1: mixed $value (array of headers)
-        if (ddtrace_call_function(fn_proxy, ZSTR_VAL(dd_inject_func), ZSTR_LEN(dd_inject_func), &retval, 2, ch,
-                                  &headers) == SUCCESS) {
-            zval_ptr_dtor(&retval);
-        } else {
-            ddtrace_log_debug("Could not inject distributed tracing headers");
-        }
-        dd_should_save_headers = true;
-        ddtrace_sandbox_end(&backup);
+    // avoiding going through our own function, directly calling curl_setopt
+    zend_execute_data *call = zend_vm_stack_push_call_frame(ZEND_CALL_TOP_FUNCTION, setopt_fn, 3, NULL);
+    ZVAL_COPY(ZEND_CALL_ARG(call, 1), ch);
+    ZVAL_COPY(ZEND_CALL_ARG(call, 2), &http_header_const->value);
+    ZVAL_COPY_VALUE(ZEND_CALL_ARG(call, 3), &headers);
 
-        dd_free_headers_arg(&headers);
-    }
+    zval ret;
+    dd_curl_setopt_handler(call, &ret);
+
+    zend_vm_stack_free_args(call);
+    zend_vm_stack_free_call_frame(call);
```

Comment:
Would this be the preferred method to call all internal functions? If so, it might be a good candidate to add as an API to ZAI functions.

Reply:
It is only a preferred way under a very specific circumstance: we _wrap_ an userland function, which is called from an internal function and that function would be anyway called, and not because of our instrumentation. Thus we don't want exception handling etc.
I'm not sure whether it's a really good ZAI candidate, as this one is only sensible and even safe to do under limited cirucmstances. I doubt that it's a good idea to have a generic abstraction hiding function for something like this (e.g. let's suppose here were a ZAI call. You see this, are inspired by a seemingly neutral sounding function name and ... use it at the wrong place - where it would be technically fine, except exceptions are not handled). I fear this would be more footgun than helpful.

Reply:
I don't think so, in general I'd just use the normal zai framework to call functions, especially internal function which possibly throw (curl_setopt does not). This is a very specific case.